cns/restserver/homeazmonitor.go

```diff
@@ -0,0 +1,168 @@
+package restserver
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/Azure/azure-container-networking/cns"
+	"github.com/Azure/azure-container-networking/cns/types"
+	"github.com/Azure/azure-container-networking/log"
+	"github.com/Azure/azure-container-networking/nmagent"
+	"github.com/patrickmn/go-cache"
+	"github.com/pkg/errors"
+)
+
+const (
+	GetHomeAzAPIName = "GetHomeAz"
+	ContextTimeOut   = 2 * time.Second
+	homeAzCacheKey   = "HomeAz"
+)
+
+type HomeAzMonitor struct {
+	nmagentClient
+	values *cache.Cache
+	// channel used as signal to end of the goroutine for populating home az cache
+	closing chan struct{}
+	// channel used as signal to block the Start() of HomeAzMonitor until the first request to retrieve home az completes
+	block                    chan struct{}
+	cacheRefreshIntervalSecs time.Duration
+}
+
+// New creates a new HomeAzMonitor object
+func New(client nmagentClient, cacheRefreshIntervalSecs time.Duration) *HomeAzMonitor {
+	return &HomeAzMonitor{
+		nmagentClient:            client,
+		cacheRefreshIntervalSecs: cacheRefreshIntervalSecs,
+		values:                   cache.New(cache.NoExpiration, cache.NoExpiration),
+		closing:                  make(chan struct{}),
+		block:                    make(chan struct{}),
+	}
+}
+
+// GetHomeAz returns home az cache value directly
+func (h *HomeAzMonitor) GetHomeAz(_ context.Context) cns.GetHomeAzResponse {
+	return h.readCacheValue()
+}
+
+// updateCacheValue updates home az cache value
+func (h *HomeAzMonitor) updateCacheValue(resp cns.GetHomeAzResponse) {
+	h.values.Set(homeAzCacheKey, resp, cache.NoExpiration)
+}
+
+// readCacheValue reads home az cache value
+func (h *HomeAzMonitor) readCacheValue() cns.GetHomeAzResponse {
+	cachedResp, _ := h.values.Get(homeAzCacheKey)
+	return cachedResp.(cns.GetHomeAzResponse)
+}
+
+// Start starts a new thread to refresh home az cache
+func (h *HomeAzMonitor) Start() {
+	go h.refresh()
+	// block until the first request to nmagent for retrieving home az completes
+	<-h.block
```

Comment:
why do need to block here?

Reply:
we want to make sure the home az cache is not empty (at least wait until the first request to retrieve home az from nmagent completed)

Reply:
what's the issue with having the empty cache when CNS gets a request from DNC? Can CNS not respond accordingly if the cache is empty indicating internal error?

Reply:
with this block we are adding latency to CNS start and I am looking to see if we can avoid that

Reply:
understand that. we can response internal error if the cache is empty